WIP: Restricted contexts draft (#2549)
circleci/circleci-docs: jekyll/_cci2/contexts.md

To rerun a job and use the context, it **must** be rerun from the Workflows page of the CircleCI application.

## Securing a Context

> CircleCI enables you to secure secrets at run time by adding security groups to contexts. Only Org Admins may add *security groups* to a new or existing context. Security groups are definied as LDAP groups or GitHub teams. After the security group is added to a context, only members of that security group may access or use the environment variables of the secure context. 

Reviewer: secure secrets -> Should we say secrets or environment variables?

Author: thanks, changed to 'secret environment variables' and 'organizaiton administrators'

The default security group is `organization` and enables any member of the org who uses CircleCI to use the context.

> ## Runnning Workflows with a Secure Context

Reviewer 2: s/Runnning/Running/

> To invoke a workflow that uses a secure context, a user must be a member of one of the security groups for the context or the workflow will fail with the status of `Unauthorized`. If you add a context to your workflow and you are **not** a member of any of the context groups, the job will fail as `Unauthorized`.

Reviewer: >the job will fail as `Unauthorized`
the workflow will fail as `Unauthorized`.

Author: thanks, fixed

**Note:** Bitbucket repositories do **not** provide an API that allows CircleCI contexts to be secured, only GitHub hosted projects or projects that use GitHub Enterprise with CircleCI include the ability to limit contexts with security groups.

### Limit a Context to a Security Group or Groups

> You must be an Org Admin to complete the following task.

Reviewer: That is not really true. Anyone can create a context but only org admins can secure them.

Author: Right, the create a context task is earlier in the document, this task is specific to securing the context

2. Click the Create Context button if you will use a new context or click the name of an existing context.
3. Click the Add Group link. The Add Groups dialog box appears.
4. Select GitHub teams or LDAP groups to add to the context and click the Add button. Use of the context is now limited to the selected groups.
5. Click Add Environment Variables to add environment variables to the context if none exist and click the Add button. Use of the environment variables for this context is now limited to members of the security groups.
6. Navigate to Organization Settings > Contexts in the CircleCI app. The security groups appear in the Security column for the context.

Only members of the selected groups may now use the context in their workflows or add or remove environment variables for the context.

## Removing Groups from Contexts

To make a context available only to the admins of the org, you may remove all of the groups associated with a context. All other users will lose access to that context.

### Adding and Removing Users from Teams and Groups

> CircleCI syncs GitHub team and LDAP groups every four hours. If a user is added or removed from a GitHub team or LDAP group, it will take up to four hours to update the CircleCI records and remove access to the context.

Reviewer: Can we remove `four` hours and say every few hours.

Author: sure

### Adding and Removing Environment Variables from Secure Contexts

Addition and deletion of environment variables from a secure context is limited to members of the context groups.

### Approving Jobs that use Secure Contexts

> Manual invocation of a job that is `type: approval` is limited to members of the context groups.

Reviewer: This is tricky. Using approval, users can switch actors. Users can't add context to approval jobs. What they can do instead is add an approval job to switch actors. Example: If Adam started the workflow and their is an approval job as part of the workflow. If Michelle, approves the job and the following workflow will use Michelle's permission.

Author: hmm, why is what I say here wrong? if a job uses a secure context and an approval job, the approval must be done by a member of the security group, right?

Reviewer: >the approval must be done by a member of the security group, right?
You can't add `context` to an approval job. `Context` can only be added to jobs that are executed.

## Deleting a Context

> If the context is secured with a group other than `organization`, you must be an org admin to complete this task.

Reviewer: oh that is not true.. Any one who is member of the group can delete the context.

Author: Ok, thanks,  changed to "If the context is secured with a group other than `organization`, you must be a member of the security group to complete this task."

To rerun a job and use the context, it **must** be rerun from the Workflows page of the CircleCI application.

## Restricting a Context

CircleCI enables you to restrict secret environment variables at run time by adding security groups to contexts. Only organization administrators may add *security groups* to a new or existing context. Security groups are definied as LDAP groups or GitHub teams. After a security group is added to a context, only members of that security group who are also CircleCI users may access or use the environment variables of the secure context. 

The default security group is `organization` and enables any member of the org who uses CircleCI to use the context.

## Running Workflows with a Restricted Context

To invoke a workflow that uses a restricted context, a user must be a member of one of the security groups for the context or the workflow will fail with the status of `Unauthorized`. If you add a context to your workflow and you are **not** a member of any of the context groups, the workflow will fail as `Unauthorized`.

> **Note:** Bitbucket repositories do **not** provide an API that allows CircleCI contexts to be restricted, only GitHub  projects include the ability to restrict contexts with security groups. Restricted Contexts are also **not** yet supported in private installations of CircleCI.

Reviewer 2: ```suggestion
**Note:** Bitbucket repositories do **not** provide an API that allows CircleCI contexts to be restricted, only GitHub  projects include the ability to restrict contexts with security groups. Restricted Contexts are also **not** yet supported in private installations of CircleCI.
```
```suggestion
**Note:** Bitbucket repositories do **not** provide an API that allows CircleCI contexts to be restricted, only GitHub projects include the ability to restrict contexts with security groups. Restricted Contexts are also **not** yet supported in private installations of CircleCI.
```

### Restrict a Context to a Security Group or Groups

You must be an Org Admin to complete the following task.

> 1. Navigate to Organization Settings > Contexts page in the CircleCI app. The list of contexts appears. The default security group is `organization` and allows all users in the org to invoke jobs with that context.

Reviewer 2: ```suggestion
1. Navigate to Organization Settings > Contexts page in the CircleCI app. The list of contexts appears. The default security group is `All members` and allows all users in the org to invoke jobs with that context.
```

2. Click the Create Context button if you will use a new context or click the name of an existing context.
3. Click the Add Group link. The Add Groups dialog box appears.
4. Select GitHub teams or LDAP groups to add to the context and click the Add button. Use of the context is now limited to the selected groups.
5. Click Add Environment Variables to add environment variables to the context if none exist and click the Add button. Use of the environment variables for this context is now limited to members of the security groups.
6. Navigate to Organization Settings > Contexts in the CircleCI app. The security groups appear in the Security column for the context.

Only members of the selected groups may now use the context in their workflows or add or remove environment variables for the context. 

### Approving Jobs that use Restricted Contexts

Adding an approval job to a workflow allows a user to manually approve the use of a restricted context. 

To restrict running of jobs that are downstream from an approval job, add a restricted context to those downstream jobs. 

> For example, if you want the execution of job C and job D restricted to a secutiry group, you need to add an approval job B before the jobs C and D to that use a context with a security group. That is, you can have four jobs in a workflow, job A can run unrestricted, the approval job B may be approved by any member, but the jobs C and D after the approval may only be executed by someone in the security group for the context used on jobs C and D. 

Reviewer 2: ```suggestion
For example, if you want the execution of job C and job D restricted to a security group, you need to add an approval job B before the jobs C and D to that use a context with a security group. That is, you can have four jobs in a workflow, job A can run unrestricted, the approval job B may be approved by any member, but the jobs C and D after the approval may only be executed by someone in the security group for the context used on jobs C and D. 
```

> If the approver of a job is not part of the restricted context, it is possible to approve the job B, however the jobs C and D in the workflow will fail as unauthorized. That is, the Approval button will appear for every user, even for users who are not part of the group with permissions for the context. When the downstream jobs fail with Unauthorized, it indicates an approval was made by a user who is not part of the security group for the downstream jobs. 

Reviewer 3: This UI appears to violate the [principle of least astonishment](https://en.wikipedia.org/wiki/Principle_of_least_astonishment). Say I'm not in the restricted list, how can I know that I should not try to approve but find someone who can? Do I need to try it, fail, and then try to get someone else to re-run it from failed for me? (And how can I find out who is in the group?)

Would you consider allowing the restricted context to work for `type: approval` jobs too, so people can model a better UX than described here? I get that the restriction has to be for using the context, otherwise people could just edit the workflow to remove the approval.

## Removing Groups from Contexts

To make a context available only to the admins of the org, you may remove all of the groups associated with a context. All other users will lose access to that context.

### Adding and Removing Users from Teams and Groups

CircleCI syncs GitHub team and LDAP groups every few hours. If a user is added or removed from a GitHub team or LDAP group, it will take up to a few hours to update the CircleCI records and remove access to the context.

### Adding and Removing Environment Variables from Restricted Contexts

Addition and deletion of environment variables from a restricted context is limited to members of the context groups.

## Deleting a Context

> If the context is restricted with a group other than `organization`, you must be a member of the security group to complete this task.

Reviewer 2: ```suggestion
If the context is restricted with a group other than `All members`, you must be a member of the security group to complete this task.
```
